tests/test_kernels.py

```diff
@@ -0,0 +1,89 @@
+"""Kernel tests."""
+
+from typing import Any
+
+import numpy as np
+import torch
+from attrs import asdict, has
+from hypothesis import given
+
+from baybe.kernels.base import Kernel
+from baybe.kernels.composite import AdditiveKernel, ProductKernel, ScaleKernel
+
+from .hypothesis_strategies.kernels import kernels
+
+# TODO: Consider deprecating these attribute names to avoid inconsistencies
+_RENAME_DICT: dict[str, str] = {
+    "base_kernels": "kernels",
+}
+"""Dictionary for resolving name differences between BayBE and GPyTorch attributes."""
+
+
+def validate_kernel_components(obj: Any, mapped: Any, **kwargs) -> None:
+    """Validate that all kernel components are translated correctly.
+
+    Args:
+        obj: An object occurring as part of a BayBE kernel.
+        mapped: The corresponding object in the translated GPyTorch kernel.
+        **kwargs: Optional kernel arguments that were passed to the GPyTorch kernel.
+    """
+    # Assert that the kernel kwargs are correctly mapped
+    # TODO: Consider introducing intermediate kernel classes or class variables
+    #   to differentiate basic kernels from composite ones
+    if isinstance(obj, Kernel) and not isinstance(
+        obj, (ScaleKernel, AdditiveKernel, ProductKernel)
+    ):
+        for k, v in kwargs.items():
+            assert torch.tensor(getattr(mapped, k)).equal(torch.tensor(v))
+
+    # Compare attribute by attribute
+    for name, component in asdict(obj, recurse=False).items():
+        # If the BayBE component is `None`, the GPyTorch component might not even
+        # exist, so we skip
+        if component is None:
+            continue
+
+        # Resolve attribute naming differences
+        mapped_name = _RENAME_DICT.get(name, name)
+
+        # If the attribute does not exist in the GPyTorch version, it must be an
+        # initial value. Because setting initial values involves going through
+        # constraint transformations on GPyTorch side (i.e., difference between
+        # `<attr>` and `raw_<attr>`), the numerical values will not be exact, so
+        # we check only for approximate matches.
+        if (mapped_component := getattr(mapped, mapped_name, None)) is None:
+            assert name.endswith("_initial_value")
+            assert np.allclose(
+                component,
+                getattr(mapped, name.removesuffix("_initial_value")).detach().numpy(),
+            )
+
+        # If the component is itself another BayBE object, recurse
+        elif has(component):
```

Comment:
Technically, this checks whether the component is itself another `attrs` object, right? Would thus change the comment to reflect that since we do not know if we might ever want to include something here that also is built via `attrs` but not a BayBE object.

Reply:
changed